pkg/disttask/framework/taskexecutor/task_executor.go

```diff
@@ -223,19 +241,56 @@ func (e *BaseTaskExecutor) Run(resource *proto.StepResource) {
 			}
 		}
 		skipBackoff = false
-		if err := e.refreshTask(); err != nil {
+		oldTask := e.task.Load()
+		failpoint.InjectCall("beforeGetTaskByIDInRun", oldTask.ID)
+		newTask, err := e.taskTable.GetTaskByID(e.ctx, oldTask.ID)
+		if err != nil {
 			if errors.Cause(err) == storage.ErrTaskNotFound {
 				return
 			}
 			e.logger.Error("refresh task failed", zap.Error(err))
 			continue
 		}
-		task := e.task.Load()
-		if task.State != proto.TaskStateRunning {
+
+		if !bytes.Equal(oldTask.Meta, newTask.Meta) {
+			e.logger.Info("task meta modified",
+				zap.String("oldStep", proto.Step2Str(oldTask.Type, oldTask.Step)),
+				zap.String("newStep", proto.Step2Str(newTask.Type, newTask.Step)))
+			// when task switch to next step, task meta might change too, but in
+			// this case step executor will be recreated, so we only notify it
```

Comment:
just double check (I didn't check the code) if the new executor is created, it will write task meta with new task step? Will the new concurrency be overwritten in this case?

Reply:
yes, the new created executor will use new task info

Reply:
I mean if there's some data race of updating concurrency and updating step.

Reply:
at this loop, before `runSubtask`, step executor is only read/write by this routine.